### ISSUE TITLE
Unified Server Task API

### ISSUE THREAD

**Issue by matthewpi** [pterodactyl/panel#4569]
There are many parts of this software that require exclusively locking parts of an individual server.  Currently very few operations actually properly lock the server, a few examples of this are server installation, suspension, and transfers.  However, there are many more operations that proceed in the background and may not be stopped or paused when the server is "locked", such as file extraction, remote file pull, backups, etc.  This poses a problem as there is not a unified way to monitor the operations on a server, and very few of these operations are able to be canceled or aborted.  This is why I am proposing a new task API for servers which should resolve many of these problems.

The Unified Server Task API would be a component of Wings that essentially acts as a per-server registry of tasks being executed on a server, tasks would include operations like remote file download, file decompression, archive extraction, backup creation, backup restore, server (re)installation, and server transfers.  The Task API would not inherently change how these tasks are ran, but how they are wired up within Wings and capabilities they should support.  Right now, very few operations in Wings support the receiving a `context`, which can be used to support operation cancellation and timeouts.  This poses a problem when initiating another task that requires waiting for or canceling all other tasks, such as a server transfer or backup restore.

The solution to this is to add a task registry on every server with the tasks being executed in a centralized manor.  Currently some tasks are tracked independently and have a complicated and messy way of tracking progress, a great example of this is [remote file downloads](https://github.com/pterodactyl/wings/blob/ff50d0e5bd158110366aceb2ae05cd2df58783e4/router/downloader/downloader.go#L108).  Remote file downloads have their own registry managed as a package level variable, this is a terrible design for multiple reasons, but is done because with the way Wings is architected makes handling stuff like this properly very complicated and hard to debug.

---

Each server would have it's own task registry linked to it and accessible directly on the server struct itself.  The registry would have two types of tasks, exclusive and background tasks.  Exclusive tasks block all other tasks and usually parts of the Server API, examples of exclusive tasks would be, suspension, (re)installation, transfers, and backup restore; only one exclusive task can be ran at a time.  These tasks require the server's power state to be locked and only controlled by the running task and all existing and future filesystem operations to be blocked while the task is running.  Background tasks are able to be ran in parallel and don't have any conditions about blocking other tasks, examples of background tasks include, remote file download, file decompression, archive extraction, and backup creation.  Background tasks are an important part of this feature as by registering them we can easily monitor progress and cancel them should the need arise (such as when deleting a server or needing to run an exclusive task).

By having a task registry for each server, we could also potentially persist parts of the server state to survive when Wings is restarted to either properly cleanup tasks or to continue their progress.  For example, if a remote download task is running and Wings restarted (assuming the task information was persisted), we could try to resume where we were before the task was aborted by checking the length of the already downloaded content and using a `Range` request to the URL where the file was being downloaded from.  Not all tasks will support that level of resumability, tasks such as backup creation and restoration get very complicated to resume unless we know their EXACT state before Wings was stopped.

Part of exclusive tasks would also be the option to wait for all background tasks to complete (or maybe even a timeout) or to cancel all other tasks immediately so the exclusive task can run.  Two different examples of this would be a server transfer and a server suspension.  When transferring a server we would want to prevent any new tasks from being scheduled, wait for the server to gracefully stop, finish any file operations, and then ultimately start transferring the server.  The behavior for a transfer differs from a suspension as a suspension is usually a forceful operation, all tasks are canceled and we attempt to gracefully stop the server with a maximum timeout where it will be forcefully killed.

**Comment by DaneEveritt** [comment]
In the interest of maintainer sanity and cleaning up an enormous backlog of stale feature requests we have moved to requiring all feature requests be discussed within [Discussions](https://github.com/orgs/pterodactyl/discussions/categories/feature-requests). Please consider re-opening this request there if you believe it is still relevant and worth consideration.

See https://github.com/pterodactyl/panel/issues/5491 for additional context.

Thanks.

<sub>_This comment is in the interest of issue tracker maintenance: it does not imply that maintainers are endorsing or dismissing your request._</sub>
